File: npm/src/handoff.ts
/** Handoff document generation, validation, and persistence. */

import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import type { State } from "./state.js";
import type { StateStore } from "./state.js";

export const WORD_LIMIT = 750;

export const REQUIRED_SECTIONS = [
  "## Code Style",
  "## Original Prompt / Current Phase",
  "## Completed Results",
  "## To-Do",
  "## Sprints / Phases for Next Context Window",
] as const;

export function countWords(text: string): number {
  let inCode = false;
  let words = 0;
  for (const line of text.split("\n")) {
    const stripped = line.trim();
    if (stripped.startsWith("```")) {
      inCode = !inCode;
      continue;
    }
    if (inCode) continue;
    const matches = line.match(/\b\w+\b/g);
    if (matches) words += matches.length;
  }
  return words;
}

export function validate(text: string): string[] {
  const problems: string[] = [];
  const wc = countWords(text);
  if (wc > WORD_LIMIT) {
    problems.push(`word count ${wc} exceeds limit of ${WORD_LIMIT}`);
  }
  for (const section of REQUIRED_SECTIONS) {
    if (!text.includes(section)) {
      problems.push(`missing required section: '${section}'`);
    }
  }
  return problems;
}

export function truncateToLimit(text: string, limit = WORD_LIMIT): string {
  if (countWords(text) <= limit) return text;
  const lines = text.split("\n");
  while (lines.length > 0 && countWords(lines.join("\n")) > limit) {
    lines.pop();
  }
  return lines.join("\n").replace(/\s+$/, "") + "\n";
}

export function writeHandoff(
  store: StateStore,
  text: string,
  label = "",
): string {
  mkdirSync(store.handoffsDir, { recursive: true });
  const stamp = new Date()
    .toISOString()
    .replace(/[-:]/g, "")
    .replace(/\..+/, "Z");
  const suffix = label ? `-${label}` : "";
  const path = join(store.handoffsDir, `handoff-${stamp}${suffix}.md`);
  writeFileSync(path, text, "utf8");
  return path;
}

export function parseHandoff(text: string): Record<string, string> {
  const sections: Record<string, string> = {};
  let currentKey: string | null = null;
  let buffer: string[] = [];
  for (const line of text.split("\n")) {
    if (line.startsWith("## ")) {
      if (currentKey !== null) {
        sections[currentKey] = buffer.join("\n").trim();
      }
      currentKey = line.slice(3).trim();
      buffer = [];
    } else if (currentKey !== null) {
      buffer.push(line);
    }
  }
  if (currentKey !== null) {
    sections[currentKey] = buffer.join("\n").trim();
  }
  return sections;
}

export function hydrateStateFromHandoff(text: string, state: State): State {
  const sections = parseHandoff(text);

  if (!state.code_style && sections["Code Style"]) {
    state.code_style = sections["Code Style"]!;
  }

  const op = sections["Original Prompt / Current Phase"];
  if (!state.original_prompt && op) {
    state.original_prompt = op.split("\n\n", 1)[0]!.trim();
  }

  const todoSection = sections["To-Do"];
  if (todoSection && state.todos.length === 0) {
    for (const line of todoSection.split("\n")) {
      const stripped = line.trim();
      if (stripped.startsWith("- ") || stripped.startsWith("* ")) {
        state.addTodo(stripped.slice(2).trim());
      }
    }
  }

  return state;
}

File: src/ccc/handoff.py
"""Handoff document generation, validation, and persistence."""

from __future__ import annotations

import re
from datetime import datetime, timezone
from pathlib import Path

from ccc.state import State, StateStore

WORD_LIMIT = 750

REQUIRED_SECTIONS = (
    "## Code Style",
    "## Original Prompt / Current Phase",
    "## Completed Results",
    "## To-Do",
    "## Sprints / Phases for Next Context Window",
)


def count_words(text: str) -> int:
    """Word count that ignores Markdown fences and code-block lines."""
    in_code = False
    words = 0
    for line in text.splitlines():
        stripped = line.strip()
        if stripped.startswith("```"):
            in_code = not in_code
            continue
        if in_code:
            continue
        words += len(re.findall(r"\b\w+\b", line))
    return words


def validate(text: str) -> list[str]:
    """Return a list of validation problems. Empty list = doc is good."""
    problems: list[str] = []
    wc = count_words(text)
    if wc > WORD_LIMIT:
        problems.append(f"word count {wc} exceeds limit of {WORD_LIMIT}")
    for section in REQUIRED_SECTIONS:
        if section not in text:
            problems.append(f"missing required section: {section!r}")
    return problems


def truncate_to_limit(text: str, limit: int = WORD_LIMIT) -> str:
    """Trim trailing lines until the document is under the word limit."""
    if count_words(text) <= limit:
        return text
    lines = text.splitlines()
    while lines and count_words("\n".join(lines)) > limit:
        lines.pop()
    return "\n".join(lines).rstrip() + "\n"


def write_handoff(store: StateStore, text: str, label: str = "") -> Path:
    """Persist a handoff document under `.ccc/handoffs/` and return the path."""
    store.handoffs_dir.mkdir(parents=True, exist_ok=True)
    stamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    suffix = f"-{label}" if label else ""
    path = store.handoffs_dir / f"handoff-{stamp}{suffix}.md"
    path.write_text(text, encoding="utf-8")
    return path


def parse_handoff(text: str) -> dict[str, str]:
    """Split a handoff document into a {section: body} map."""
    sections: dict[str, str] = {}
    current_key: str | None = None
    buffer: list[str] = []
    for line in text.splitlines():
        if line.startswith("## "):
            if current_key is not None:
                sections[current_key] = "\n".join(buffer).strip()
            current_key = line[3:].strip()
            buffer = []
        elif current_key is not None:
            buffer.append(line)
    if current_key is not None:
        sections[current_key] = "\n".join(buffer).strip()
    return sections


def hydrate_state_from_handoff(text: str, state: State) -> State:
    """Reload a State from a previously-written handoff document.

    Only fills empty fields; never overwrites live state.
    """
    sections = parse_handoff(text)

    if not state.code_style:
        state.code_style = sections.get("Code Style", "")

    op_section = sections.get("Original Prompt / Current Phase", "")
    if not state.original_prompt and op_section:
        first_para = op_section.split("\n\n", 1)[0].strip()
        state.original_prompt = first_para

    todo_section = sections.get("To-Do", "")
    if todo_section and not state.todos:
        for line in todo_section.splitlines():
            stripped = line.strip()
            if stripped.startswith(("- ", "* ")):
                state.add_todo(stripped[2:].strip())

    return state
